beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go

```diff
@@ -197,86 +194,145 @@ func (vs *Server) BuildBlockParallel(ctx context.Context, sBlk interfaces.Signed
 	return nil
 }
 
-// ProposeBeaconBlock is called by a proposer during its assigned slot to create a block in an attempt
-// to get it processed by the beacon node as the canonical head.
+// ProposeBeaconBlock handles the proposal of beacon blocks.
 func (vs *Server) ProposeBeaconBlock(ctx context.Context, req *ethpb.GenericSignedBeaconBlock) (*ethpb.ProposeResponse, error) {
 	ctx, span := trace.StartSpan(ctx, "ProposerServer.ProposeBeaconBlock")
 	defer span.End()
 
-	blk, err := blocks.NewSignedBeaconBlock(req.Block)
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	block, err := blocks.NewSignedBeaconBlock(req.Block)
 	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, "%s: %v", CouldNotDecodeBlock, err)
+		return nil, errors.Wrap(err, "decode block failed")
 	}
 
-	unblinder, err := newUnblinder(blk, vs.BlockBuilder)
+	var sidecars []*ethpb.BlobSidecar
+	if block.IsBlinded() {
+		block, sidecars, err = vs.handleBlindedBlock(ctx, block)
+	} else {
+		sidecars, err = vs.handleUnblindedBlock(block, req)
+	}
 	if err != nil {
-		return nil, errors.Wrap(err, "could not create unblinder")
+		return nil, err
+	}
+
+	root, err := block.Block().HashTreeRoot()
```

Comment:
Do we need to assert that `Block() != nil`?

Reply:
I thought about this, but given we called `NewSignedBeaconBlock` before, we should be fine
`block, err := blocks.NewSignedBeaconBlock(req.Block)` should not return a nil `block` or `block.Block()`